### PR TITLE
Improve collection of dataset usages and citation on occurrence downloads 

### PR DESCRIPTION
**Pull request by aalbatross**
List of changes:
1. Added new webservice endpoint: HTTP POST (occurrence/download/{downloadkey}/datasets}.
2. Removed webservice endpoint : DatasetOccurrenceDownloadUsageResource:create
3. Added and removed endpoints in registry-ws-client accordingly.
4. Added integration test to test new endpoint.

### CODE REVIEW

File: registry-ws-client/src/main/java/org/gbif/registry/ws/client/DatasetOccurrenceDownloadUsageWsClient.java [hunk 1]
```diff
@@ -21,19 +20,15 @@
 public class DatasetOccurrenceDownloadUsageWsClient extends BaseWsGetClient<Download, String> implements
   DatasetOccurrenceDownloadUsageService {
 
+  private static final String DATASET = "dataset";
```
Comment (fmendezh) [inline registry-ws-client/src/main/java/org/gbif/registry/ws/client/DatasetOccurrenceDownloadUsageWsClient.java]: DATASET_PATH?

File: registry-ws-client/src/main/java/org/gbif/registry/ws/client/OccurrenceDownloadWsClient.java [hunk 2]
```diff
@@ -104,4 +104,9 @@ public void create(Download occurrenceDownload) {
     Optional.ofNullable(datasetKey).ifPresent( dk -> params.add("datasetKey", dk.toString()));
     return get(GenericTypes.DOWNLOADS_STATS_TYPE, null, params ,null,path);
   }
+
+  @Override
+  public void createUsages(String downloadKey, Map<UUID, Long> datasetCitations) {
+    post(datasetCitations,downloadKey+"/datasets");
```
Comment (fmendezh) [inline registry-ws-client/src/main/java/org/gbif/registry/ws/client/OccurrenceDownloadWsClient.java]: use spaces around + and after each ,
OR use the varargs post(datasetCitations, downloadKey, "datasets");

File: registry-ws/src/main/java/org/gbif/registry/persistence/mapper/DatasetOccurrenceDownloadMapper.java [hunk 3]
```diff
@@ -16,23 +16,23 @@
...
-
   List<DatasetOccurrenceDownloadUsage> listByDataset(@Param("datasetKey") UUID datasetKey,
     @Nullable @Param("page") Pageable page);
 
   int countByDataset(@Param("datasetKey") UUID datasetKey);
 
   List<DatasetOccurrenceDownloadUsage> listByDownload(@Param("downloadKey") String downloadKey,
                                                      @Nullable @Param("page") Pageable page);
+  
+  void createUsages(@Param("downloadkey")String downloadkey,@Param("citationMap")Map<UUID,Long> downloadDataset);
```
Comment (fmendezh) [inline registry-ws/src/main/java/org/gbif/registry/persistence/mapper/DatasetOccurrenceDownloadMapper.java]: uses blanks after ) and ,

File: registry-ws/src/main/java/org/gbif/registry/ws/resources/DatasetOccurrenceDownloadUsageResource.java [hunk 4]
```diff
@@ -27,10 +21,6 @@
 
 import com.google.inject.Inject;
 import com.google.inject.Singleton;
-import org.apache.bval.guice.Validate;
-import org.mybatis.guice.transactional.Transactional;
-
-import static org.gbif.registry.ws.security.UserRoles.ADMIN_ROLE;
 import static org.gbif.registry.ws.util.DownloadSecurityUtils.clearSensitiveData;
 
 /**
```
Comment (fmendezh) [inline registry-ws/src/main/java/org/gbif/registry/ws/resources/DatasetOccurrenceDownloadUsageResource.java]: PrePersist is not used anymore
securityContext can be removed

File: registry-ws/src/main/java/org/gbif/registry/ws/resources/OccurrenceDownloadResource.java [hunk 5]
```diff
@@ -168,6 +168,16 @@ public void update(Download download) {
...
     throw new NotFoundException();
   }
+  
+  @POST
+  @Path("{key}/datasets")
+  @Transactional
+  @Validate(groups = {PrePersist.class, Default.class})
+  @RolesAllowed(ADMIN_ROLE)
+  @Override
+  public void createUsages(@PathParam("key") String downloadkey,@Valid @NotNull Map<UUID,Long> datasetCitations) {
```
Comment (fmendezh) [inline registry-ws/src/main/java/org/gbif/registry/ws/resources/OccurrenceDownloadResource.java]: add blank space after ,

File: registry-ws/src/main/resources/org/gbif/registry/persistence/mapper/DatasetOccurrenceDownloadMapper.xml [hunk 6]
```diff
@@ -53,5 +41,20 @@
     FROM dataset_occurrence_download
     WHERE dataset_key = #{datasetKey,jdbcType=OTHER}
   </select>
+  
+  <insert id="createUsages">
+	INSERT INTO dataset_occurrence_download (
+	WITH dataset_usages(dataset_key, number_records) AS
+	((VALUES
+	<foreach item="value" index="key" collection="citationMap"
```
Comment (fmendezh) [inline registry-ws/src/main/resources/org/gbif/registry/persistence/mapper/DatasetOccurrenceDownloadMapper.xml]: foreach in line looks more readable: <foreach item="value" index="key" collection="citationMap" open="(" separator="),(" close=")">

File: registry-ws/src/main/resources/org/gbif/registry/persistence/mapper/DatasetOccurrenceDownloadMapper.xml [hunk 7]
```diff
@@ -53,5 +41,20 @@
...
+		open="(" separator="),(" close=")">
+		cast(#{key} as uuid),#{value}
+	</foreach>
+	))
+	SELECT #{downloadkey} as download_key, dataset.key, number_records,
```
Comment (fmendezh) [inline registry-ws/src/main/resources/org/gbif/registry/persistence/mapper/DatasetOccurrenceDownloadMapper.xml]: SELECT  in 1 line 
SELECT #{downloadkey} as download_key, dataset.key, number_records, title, doi, citation

File: registry-ws/src/main/java/org/gbif/registry/ws/resources/DatasetOccurrenceDownloadUsageResource.java (hunk 4 above)
Comment (aalbatross) [reply registry-ws/src/main/java/org/gbif/registry/ws/resources/DatasetOccurrenceDownloadUsageResource.java]: can't remove, listByDataset uses it.

File: registry-ws/src/main/java/org/gbif/registry/ws/resources/OccurrenceDownloadResource.java [hunk 8]
```diff
@@ -168,6 +168,16 @@ public void update(Download download) {
     }
     throw new NotFoundException();
```
Comment (fmendezh) [inline registry-ws/src/main/java/org/gbif/registry/ws/resources/OccurrenceDownloadResource.java]: remove unused private SecurityContext securityContext; and related import

Comment (aalbatross) [reply registry-ws/src/main/java/org/gbif/registry/ws/resources/OccurrenceDownloadResource.java]: securityContext is used in the same java file can't remove.

**Pull request merged**